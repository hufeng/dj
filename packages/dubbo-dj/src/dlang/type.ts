import { IType, TDataType } from '../types'
import { DubboDep } from './lang-deps'
import { DubboEntity } from './lang-entity'
import { DubboEnum } from './lang-enum'

export interface IProject {
  name: string
  version: string
  description: string
}

// ~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~~~
export function universalType(dep: DubboDep, type: TDataType) {
  if (type instanceof DubboEntity) {
    dep.add({
      fromModule: type.fullName,
      importModule: type.shortName,
    })
    return {
      tsType: type.infName,
      javaType: `((v:any) => v.__fields2java())`,
    }
  } else if (type instanceof DubboEnum) {
    dep.add({
      fromModule: type.fullName,
      importModule: type.shortName,
    })
    return {
      tsType: type.shortName,
      javaType: `((v:any) => java.enum('${type.fullName}', ${type.shortName}[v]))`,
    }
  } else {
    return type(dep)
  }
}

// ~~~~~~~~~ Boolean && boolean ~~~~~~~~~~~~~~~
export function Boolean(): IType {
  return {
    tsType: 'boolean',
    javaType: 'java.Boolean',
  }
}

export function boolean(): IType {
  return {
    tsType: 'boolean',
    javaType: 'java.boolean',
  }
}

// ~~~~~~~~~ each number type ~~~~~~~~~~~~~~
export function Integer(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Integer',
  }
}

export function int(): IType {
  return {
    tsType: 'number',
    javaType: 'java.int',
  }
}

export function short(): IType {
  return {
    tsType: 'number',
    javaType: 'java.short',
  }
}

export function Short(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Short',
  }
}

export function byte(): IType {
  return {
    tsType: 'number',
    javaType: 'java.byte',
  }
}

export function Byte(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Byte',
  }
}

export function long(): IType {
  return {
    tsType: 'number',
    javaType: 'java.long',
  }
}

export function Long(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Long',
  }
}

export function double(): IType {
  return {
    tsType: 'number',
    javaType: 'java.double',
  }
}

export function Double(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Double',
  }
}

export function float(): IType {
  return {
    tsType: 'number',
    javaType: 'java.float',
  }
}

export function Float(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Float',
  }
}

// ~~~~~~~~~~~~~~ String ~~~~~~~~~~~~~~~~
export function String(): IType {
  return {
    tsType: 'string',
    javaType: 'java.String',
  }
}

export function char(): IType {
  return {
    tsType: 'string',
    javaType: 'java.char',
  }
}

export function Char(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Char',
  }
}

export function chars(): IType {
  return {
    tsType: 'string',
    javaType: 'java.chars',
  }
}

export function Character(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Character',
  }
}

// ~~~~~~~~~~~~~~~~~~~~~ container ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export function List(
  type: TDataType
): (() => IType) | ((dep: DubboDep) => IType) {
  if (type instanceof DubboEntity) {
    const { infName } = type
    return (dep: DubboDep) => {
      dep.add({
        fromModule: type.fullName,
        importModule: type.infName,
      })
      return {
        tsType: `Array<${infName}>`,
        javaType: `(function withList(val:Array<any>=[]) {return java.List(val.map(v => v.__fields2java()))})`,
      }
    }
  }

  if (type instanceof DubboEnum) {
    const { shortName, fullName } = type
    return (dep: DubboDep) => {
      dep.add({
        fromModule: type.fullName,
        importModule: type.shortName,
      })
      return {
        tsType: `Array<${shortName}>`,
        javaType: `(function withList(val:Array<any>=[]) {return java.List(val.map(v => java.enum('${fullName}')))})})`,
      }
    }
  }

  return (dep: DubboDep) => {
    const t = type(dep)
    return {
      tsType: `Array<${t.tsType}>`,
      javaType: `(function withList(val:Array<any>=[]){return java.List(val.map(v => ${t.javaType}(v)))})`,
    }
  }
}

/*
export function Set(type: Entity | LangEnum | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Set',
    generic: [type],
  })
}

export function Collection(type: Entity | LangEnum | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Collection',
    generic: [type],
  })
}

export function Iterator(type: Entity | LangEnum | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Iterator',
    generic: [type],
  })
}

export function Enumeration(type: Entity | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Iterator',
    generic: [type],
  })
}

export function HashMap(
  leftType: Entity | (() => IType),
  rightType: Entity | LangEnum | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.HashMap',
    generic: [leftType, rightType],
  })
}

export function Map(
  leftType: Entity | (() => IType),
  rightType: Entity | LangEnum | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.Map',
    generic: [leftType, rightType],
  })
}

export function Dictionary(
  leftType: Entity | (() => IType),
  rightType: Entity | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.Dictionary',
    generic: [leftType, rightType],
  })
}
*/

export function Currency(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Currency',
  }
}
