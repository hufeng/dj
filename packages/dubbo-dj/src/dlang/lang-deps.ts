export type TFromModule = string
export interface IDepValue {
  is3rdLib: boolean
  defaultModule: string
  noneDefaultModules: Set<string>
}
export interface IDepAdd {
  fromModule: string
  importModule: string
  isDefaultModule?: boolean
  is3rdLib?: boolean
}

export class DubboDep {
  private readonly depMap: Map<TFromModule, IDepValue>

  constructor() {
    this.depMap = new Map<TFromModule, IDepValue>()
  }

  add({
    fromModule,
    importModule,
    isDefaultModule = false,
    is3rdLib = false,
  }: IDepAdd) {
    if (this.depMap.has(fromModule)) {
      const moduleValue = this.depMap.get(fromModule) || {
        defaultModule: '',
        noneDefaultModules: new Set(),
      }
      if (isDefaultModule) {
        moduleValue.defaultModule = importModule
      } else {
        moduleValue.noneDefaultModules.add(importModule)
      }
    } else {
      if (isDefaultModule) {
        this.depMap.set(fromModule, {
          is3rdLib,
          defaultModule: importModule,
          noneDefaultModules: new Set<string>(),
        })
      } else {
        this.depMap.set(fromModule, {
          is3rdLib,
          defaultModule: '',
          noneDefaultModules: new Set([importModule]),
        })
      }
    }
  }

  getDepMap() {
    return this.depMap
  }
}
