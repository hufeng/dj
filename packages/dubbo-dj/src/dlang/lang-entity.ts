import Lang from './lang'
import { universalType } from './type'
import { IDubboEntityField, IType, TDataType } from '../types'

// ~~~~~~~~~~~~~~~~~~ model~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEntity extends Lang {
  fields: Array<IDubboEntityField>

  constructor(fullName: string, comment?: string) {
    super(fullName, comment)
    this.fields = []
    this.deps.add({
      fromModule: 'js-to-java',
      isDefaultModule: true,
      importModule: 'java',
      is3rdModule: true,
    })
  }

  field(name: string, type: IType, comment?: string) {
    this.fields.push({
      comment: this.wrapComment(comment),
      name,
      type,
    })
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.comment,
      fullName: this.fullName,
      shortName: this.shortName,
      infName: this.infName,
      fields: this.fields,
      imports: this.imports(this.deps),
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class EntityBuilder {
  private readonly entity: DubboEntity

  constructor(fullName: string, comment?: string) {
    this.entity = new DubboEntity(fullName, comment)
  }

  field(name: string, type: TDataType, comment?: string) {
    const t = universalType(this.entity.deps, type)
    this.entity.field(name, t, comment)
    return this
  }

  ok() {
    return this.entity
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~
export function entity(fullName: string, comment?: string) {
  return new EntityBuilder(fullName, comment)
}
