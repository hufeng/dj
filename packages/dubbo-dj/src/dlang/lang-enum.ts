import Lang from './lang'
import { IDubboEnumFiled } from '../types'

// ~~~~~~~~~~~~~~~~~~ model~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEnum extends Lang {
  fields: Array<IDubboEnumFiled> = []

  constructor(fullName: string, comment?: string) {
    super(fullName, comment)
  }

  field(name: string, val: number | string, comment?: string) {
    this.fields.push({
      name,
      val,
      comment: this.wrapComment(comment),
    })
    return this
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.comment,
      fullName: this.fullName,
      fields: this.fields,
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEnumDSL {
  private readonly enumeration: DubboEnum

  constructor(fullName: string, comment: string = '') {
    this.enumeration = new DubboEnum(fullName, comment)
  }

  /**
   * 枚举属性字段
   * @param name 属性名称
   * @param val
   * @param comment
   */
  field(name: string, val: number | string, comment?: string) {
    this.enumeration.field(name, val, comment)
    return this
  }

  /**
   * 完成Enum DSL的定义
   */
  ok() {
    return this.enumeration
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~
export function enumeration(cls: string, comment?: string) {
  return new DubboEnumDSL(cls, comment)
}
