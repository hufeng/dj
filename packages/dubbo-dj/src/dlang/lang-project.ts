import { IProject } from './type'

export class DubboProject {
  private readonly meta: IProject = { name: '', version: '', description: '' }

  name(name: string) {
    this.meta.name = name
    return this
  }

  version(version: string) {
    this.meta.version = version
    return this
  }

  description(description: string) {
    this.meta.description = description
    return this
  }

  ok() {
    return this.meta
  }
}

export const project = new DubboProject()
