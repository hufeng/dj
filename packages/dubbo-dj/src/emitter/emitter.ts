import debug from 'debug'
import fsx from 'fs-extra'
import { fmt } from '../common'

const FILE_EXT = {
  ts: 'ts',
  go: 'go',
  java: 'java',
}
const log = debug(`dj:emitter：say ~`)

export default abstract class Emitter {
  protected readonly genFilePath: string
  private readonly fullName: string

  protected constructor(
    fullName: string,
    lang: 'ts' | 'go' | 'java' = 'ts',
    baseDir: string = './dubbo'
  ) {
    this.fullName = fullName
    const fileExt = FILE_EXT[lang]
    const pkgDir = this.dotPath(fullName)
    this.genFilePath = `${baseDir}/${lang}/${pkgDir}.${fileExt}`
  }

  dotPath(cls: string) {
    return cls.replace(/\./g, '/')
  }

  // write to => ./dubbo/ts/**/*.ts
  async writeCode(): Promise<void> {
    try {
      await fsx.ensureFile(this.genFilePath)
      await fsx.writeFile(this.genFilePath, this.prettyCode)
      log('create %s => %s was successfully', this.fullName, this.genFilePath)
    } catch (err) {
      log('write file error %s', err)
      console.error(err)
    }
  }

  get prettyCode(): string {
    return fmt(this.code)
  }

  /**
   * 子类只需要实现该方法即可
   */
  abstract get code(): string
}
