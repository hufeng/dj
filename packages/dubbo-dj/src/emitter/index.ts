import DubboEnumEmitter from './emitter-enum'
import DubboEntityEmitter from './emitter-entity'
import DubboServiceProviderEmitter from './emitter-provider-service'
import DubboAbstractServiceEmitter from './emitter-abstract-service'
import DubboConsumerService from './emitter-consumer-service'

export {
  DubboEnumEmitter,
  DubboEntityEmitter,
  DubboConsumerService,
  DubboServiceProviderEmitter,
  DubboAbstractServiceEmitter,
}
