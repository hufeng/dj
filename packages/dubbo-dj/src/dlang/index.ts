import { entity } from './lang-entity'
import { service, f } from './lang-service'
import { enumeration } from './lang-enum'
import {
  Integer,
  String,
  Boolean,
  List,
  boolean,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  char,
  Char,
  chars,
  Character,
  Currency,
} from './type'

export const t = {
  Boolean,
  boolean,
  Integer,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  String,
  char,
  Char,
  chars,
  Character,
  List,
  Set,
  Currency,
}

export { entity, service, enumeration, f }
