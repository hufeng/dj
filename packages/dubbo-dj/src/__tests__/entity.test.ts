import * as t from '../dlang/type'
import { entity, enumeration } from '../dlang'
import { DubboEntityEmitter } from '../emitter'

const color = enumeration(
  'org.apache.dubbo.domain.enum.Color',
  '颜色类型的枚举'
)
  .field('RED', 0)
  .field('GREEN', 1)
  .field('BLUE', 2)
  .ok()

const addr = entity('org.apache.dubbo.domain.Address')
  .field('id', t.Integer, '地址id')
  .field('province', t.String, '省')
  .field('city', t.String, '市')
  .field('region', t.String, '区')
  .ok()

describe('e test suite', () => {
  it('test basic e', () => {
    const user = entity('org.apache.dubbo.domain.User')
      .field('id', t.Integer, '用户id')
      .field('name', t.String, '用户名')
      .field('email', t.String, '邮箱')
      .field('color', color)
      .ok()
    expect(user.toJSON).toMatchSnapshot()
    expect(
      new DubboEntityEmitter(user, 'ts', './').prettyCode
    ).toMatchSnapshot()
  })

  it('test first-level list', () => {
    const user = entity('org.apache.dubbo.domain.User')
      .field('id', t.Integer, '用户id')
      .field('name', t.String, '用户名')
      .field('likes', t.List(t.String), '爱好')
      .field('addrs', t.List(addr))
      .ok()
    expect(user.toJSON).toMatchSnapshot()
    expect(
      new DubboEntityEmitter(user, 'ts', './').prettyCode
    ).toMatchSnapshot()
  })

  it('test nested generic', () => {
    const user = entity('org.apache.dubbo.domain.User')
      .field('email', t.List(t.List(t.String)))
      .field('addrs', t.List(t.List(addr)))
      .ok()
    expect(user.toJSON).toMatchSnapshot()
    expect(
      new DubboEntityEmitter(user, 'ts', './').prettyCode
    ).toMatchSnapshot()
  })
})
