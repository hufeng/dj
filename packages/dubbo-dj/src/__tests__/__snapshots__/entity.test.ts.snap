// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e test suite test basic e 1`] = `
Object {
  "comment": "",
  "fields": Array [
    Object {
      "comment": "
    /**
    * 用户id
    */",
      "name": "id",
      "type": Object {
        "javaType": "java.Integer",
        "tsType": "number",
      },
    },
    Object {
      "comment": "
    /**
    * 用户名
    */",
      "name": "name",
      "type": Object {
        "javaType": "java.String",
        "tsType": "string",
      },
    },
    Object {
      "comment": "
    /**
    * 邮箱
    */",
      "name": "email",
      "type": Object {
        "javaType": "java.String",
        "tsType": "string",
      },
    },
    Object {
      "comment": "",
      "name": "color",
      "type": Object {
        "javaType": "((v:any) => java.enum('org.apache.dubbo.enum.Color', Color[v]))",
        "tsType": "Color",
      },
    },
  ],
  "fullName": "org.apache.dubbo.domain.User",
  "imports": "import  java from 'js-to-java';import {Color}from 'org.apache.dubbo.enum.Color';",
  "infName": "IUser",
  "mott": "
      /**
       * auto generated by dubbo dj
       * ~~~ 💗 machine coding 💗 ~~~
       */
    ",
  "shortName": "User",
}
`;

exports[`e test suite test basic e 2`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ 💗 machine coding 💗 ~~~
 */

import java from 'js-to-java'
import { Color } from 'org.apache.dubbo.enum.Color'

export interface IUser {
  /**
   * 用户id
   */ id: number

  /**
   * 用户名
   */ name: string

  /**
   * 邮箱
   */ email: string
  color: Color
}

export default class User {
  /**
   * 用户id
   */ id: number

  /**
   * 用户名
   */ name: string

  /**
   * 邮箱
   */ email: string
  color: Color

  constructor(props: IUser) {
    this.id = props.id
    this.name = props.name
    this.email = props.email
    this.color = props.color
  }

  __fields2java() {
    return java('org.apache.dubbo.domain.User', {
      id: java.Integer(this.id),
      name: java.String(this.name),
      email: java.String(this.email),
      color: ((v: any) => java.enum('org.apache.dubbo.enum.Color', Color[v]))(
        this.color
      ),
    })
  }
}
"
`;

exports[`e test suite test first-level list 1`] = `
Object {
  "comment": "",
  "fields": Array [
    Object {
      "comment": "
    /**
    * 用户id
    */",
      "name": "id",
      "type": Object {
        "javaType": "java.Integer",
        "tsType": "number",
      },
    },
    Object {
      "comment": "
    /**
    * 用户名
    */",
      "name": "name",
      "type": Object {
        "javaType": "java.String",
        "tsType": "string",
      },
    },
    Object {
      "comment": "
    /**
    * 爱好
    */",
      "name": "likes",
      "type": Object {
        "javaType": "(function withList(val:Array<any>=[]){return java.List(val.map(v => java.String(v)))})",
        "tsType": "Array<string>",
      },
    },
    Object {
      "comment": "",
      "name": "addrs",
      "type": Object {
        "javaType": "(function withList(val:Array<any>=[]) {return java.List(val.map(v => v.__fields2java()))})",
        "tsType": "Array<IAddr>",
      },
    },
  ],
  "fullName": "org.apache.dubbo.domain.User",
  "imports": "import  java from 'js-to-java';import {IAddr}from 'org.apache.dubbo.entity.Addr';",
  "infName": "IUser",
  "mott": "
      /**
       * auto generated by dubbo dj
       * ~~~ 💗 machine coding 💗 ~~~
       */
    ",
  "shortName": "User",
}
`;

exports[`e test suite test first-level list 2`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ 💗 machine coding 💗 ~~~
 */

import java from 'js-to-java'
import { IAddr } from 'org.apache.dubbo.entity.Addr'

export interface IUser {
  /**
   * 用户id
   */ id: number

  /**
   * 用户名
   */ name: string

  /**
   * 爱好
   */ likes: Array<string>
  addrs: Array<IAddr>
}

export default class User {
  /**
   * 用户id
   */ id: number

  /**
   * 用户名
   */ name: string

  /**
   * 爱好
   */ likes: Array<string>
  addrs: Array<IAddr>

  constructor(props: IUser) {
    this.id = props.id
    this.name = props.name
    this.likes = props.likes
    this.addrs = props.addrs
  }

  __fields2java() {
    return java('org.apache.dubbo.domain.User', {
      id: java.Integer(this.id),
      name: java.String(this.name),
      likes: (function withList(val: Array<any> = []) {
        return java.List(val.map((v) => java.String(v)))
      })(this.likes),
      addrs: (function withList(val: Array<any> = []) {
        return java.List(val.map((v) => v.__fields2java()))
      })(this.addrs),
    })
  }
}
"
`;

exports[`e test suite test nested generic 1`] = `
Object {
  "comment": "",
  "fields": Array [
    Object {
      "comment": "",
      "name": "email",
      "type": Object {
        "javaType": "(function withList(val:Array<any>=[]){return java.List(val.map(v => (function withList(val:Array<any>=[]){return java.List(val.map(v => java.String(v)))})(v)))})",
        "tsType": "Array<Array<string>>",
      },
    },
    Object {
      "comment": "",
      "name": "addrs",
      "type": Object {
        "javaType": "(function withList(val:Array<any>=[]){return java.List(val.map(v => (function withList(val:Array<any>=[]) {return java.List(val.map(v => v.__fields2java()))})(v)))})",
        "tsType": "Array<Array<IAddr>>",
      },
    },
  ],
  "fullName": "org.apache.dubbo.domain.User",
  "imports": "import  java from 'js-to-java';import {IAddr}from 'org.apache.dubbo.entity.Addr';",
  "infName": "IUser",
  "mott": "
      /**
       * auto generated by dubbo dj
       * ~~~ 💗 machine coding 💗 ~~~
       */
    ",
  "shortName": "User",
}
`;

exports[`e test suite test nested generic 2`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ 💗 machine coding 💗 ~~~
 */

import java from 'js-to-java'
import { IAddr } from 'org.apache.dubbo.entity.Addr'

export interface IUser {
  email: Array<Array<string>>
  addrs: Array<Array<IAddr>>
}

export default class User {
  email: Array<Array<string>>
  addrs: Array<Array<IAddr>>

  constructor(props: IUser) {
    this.email = props.email
    this.addrs = props.addrs
  }

  __fields2java() {
    return java('org.apache.dubbo.domain.User', {
      email: (function withList(val: Array<any> = []) {
        return java.List(
          val.map((v) =>
            (function withList(val: Array<any> = []) {
              return java.List(val.map((v) => java.String(v)))
            })(v)
          )
        )
      })(this.email),
      addrs: (function withList(val: Array<any> = []) {
        return java.List(
          val.map((v) =>
            (function withList(val: Array<any> = []) {
              return java.List(val.map((v) => v.__fields2java()))
            })(v)
          )
        )
      })(this.addrs),
    })
  }
}
"
`;
